Build failure: bcachefs-tools fails to cross-compile

This is `nixos-unstable`.  612f97239e2cc474c13c9dafa0df378058c5ad8d

Trying to build bcachefs-tools` for an aarch64-linux from x86_64-linux produces:

```
bcachefs-tools-aarch64-unknown-linux-gnu>    Compiling log v0.4.20
bcachefs-tools-aarch64-unknown-linux-gnu>     [AR]     libbcachefs.a
bcachefs-tools-aarch64-unknown-linux-gnu> /nix/store/9vafkkic27k7m4934fpawl6yip3a6k4h-bash-5.2-p21/bin/bash: line 1: ar: command not found
bcachefs-tools-aarch64-unknown-linux-gnu> make: *** [Makefile:188: libbcachefs.a] Error 127
bcachefs-tools-aarch64-unknown-linux-gnu> make: *** Waiting for unfinished jobs....
bcachefs-tools-aarch64-unknown-linux-gnu>    Compiling regex v1.10.2
```

I think `ar` comes from binutils. Adding binutils to nativeBuildInputs wasn't enough to fix it.

> I was able to fix this partially with this diff:
> 
> ```diff
> diff --git a/pkgs/by-name/bc/bcachefs-tools/package.nix b/pkgs/by-name/bc/bcachefs-tools/package.nix
> index 14f668bf5b67..c3b36b15de81 100644
> --- a/pkgs/by-name/bc/bcachefs-tools/package.nix
> +++ b/pkgs/by-name/bc/bcachefs-tools/package.nix
> @@ -6,6 +6,7 @@
>    libuuid,
>    libsodium,
>    keyutils,
> +  binutils,
>    liburcu,
>    zlib,
>    libaio,
> @@ -44,7 +45,7 @@ stdenv.mkDerivation (finalAttrs: {
>      rustPlatform.cargoSetupHook
>      rustPlatform.bindgenHook
>      makeWrapper
> -  ];
> +  ] ++ lib.optional stdenv.isAarch64 binutils.bintools;
>  
>    buildInputs = [
>      libaio
> ```
> 
> That fixed the issue but seems like our libraries are incompatible?
> 
> ```
>        >           /nix/store/wwfrj9kvfi14xclc38qfwm71ah6aawdh-binutils-2.41/bin/ld: cannot find -ludev: No such file or directory
>        >           /nix/store/wwfrj9kvfi14xclc38qfwm71ah6aawdh-binutils-2.41/bin/ld: skipping incompatible /nix/store/wnl1c1dbg60i00jqpxapfd70bhm3mgq3-systemd-minimal-libs-aarch64-unknown-linux-gnu-255.6/lib/libudev.so when searching for -ludev
>        >           /nix/store/wwfrj9kvfi14xclc38qfwm71ah6aawdh-binutils-2.41/bin/ld: skipping incompatible /nix/store/sr2520ilxww86sqscwsj7cynh38zyvxf-keyutils-aarch64-unknown-linux-gnu-1.6.3-lib/lib/libkeyutils.so when searching for -lkeyutils
>        >           /nix/store/wwfrj9kvfi14xclc38qfwm71ah6aawdh-binutils-2.41/bin/ld: skipping incompatible /nix/store/sr2520ilxww86sqscwsj7cynh38zyvxf-keyutils-aarch64-unknown-linux-gnu-1.6.3-lib/lib/libkeyutils.a when searching for -lkeyutils
>        >           /nix/store/wwfrj9kvfi14xclc38qfwm71ah6aawdh-binutils-2.41/bin/ld: cannot find -lkeyutils: No such file or directory
>        >           /nix/store/wwfrj9kvfi14xclc38qfwm71ah6aawdh-binutils-2.41/bin/ld: skipping incompatible /nix/store/sr2520ilxww86sqscwsj7cynh38zyvxf-keyutils-aarch64-unknown-linux-gnu-1.6.3-lib/lib/libkeyutils.so when searching for -lkeyutils
>        >           /nix/store/wwfrj9kvfi14xclc38qfwm71ah6aawdh-binutils-2.41/bin/ld: cannot find -laio: No such file or directory
>        >           /nix/store/wwfrj9kvfi14xclc38qfwm71ah6aawdh-binutils-2.41/bin/ld: skipping incompatible /nix/store/wnl1c1dbg60i00jqpxapfd70bhm3mgq3-systemd-minimal-libs-aarch64-unknown-linux-gnu-255.6/lib/libudev.so when searching for -ludev
>        >           /nix/store/wwfrj9kvfi14xclc38qfwm71ah6aawdh-binutils-2.41/bin/ld: cannot find -ludev: No such file or directory
>        >           /nix/store/wwfrj9kvfi14xclc38qfwm71ah6aawdh-binutils-2.41/bin/ld: skipping incompatible /nix/store/wnl1c1dbg60i00jqpxapfd70bhm3mgq3-systemd-minimal-libs-aarch64-unknown-linux-gnu-255.6/lib/libudev.so when searching for -ludev
>        >           /nix/store/wwfrj9kvfi14xclc38qfwm71ah6aawdh-binutils-2.41/bin/ld: skipping incompatible /nix/store/8rrs7z9q3ia5i5wpxfmiyh9x6ibm2ab3-liburcu-aarch64-unknown-linux-gnu-0.14.0/lib/liburcu-common.so when searching for -lurcu-common
>        >           /nix/store/wwfrj9kvfi14xclc38qfwm71ah6aawdh-binutils-2.41/bin/ld: cannot find -lurcu-common: No such file or directory
>        >           /nix/store/wwfrj9kvfi14xclc38qfwm71ah6aawdh-binutils-2.41/bin/ld: skipping incompatible /nix/store/8rrs7z9q3ia5i5wpxfmiyh9x6ibm2ab3-liburcu-aarch64-unknown-linux-gnu-0.14.0/lib/liburcu-common.so when searching for -lurcu-common
>        >           /nix/store/wwfrj9kvfi14xclc38qfwm71ah6aawdh-binutils-2.41/bin/ld: skipping incompatible /nix/store/8rrs7z9q3ia5i5wpxfmiyh9x6ibm2ab3-liburcu-aarch64-unknown-linux-gnu-0.14.0/lib/liburcu.so when searching for -lurcu
>        >           /nix/store/wwfrj9kvfi14xclc38qfwm71ah6aawdh-binutils-2.41/bin/ld: cannot find -lurcu: No such file or directory
>        >           /nix/store/wwfrj9kvfi14xclc38qfwm71ah6aawdh-binutils-2.41/bin/ld: skipping incompatible /nix/store/8rrs7z9q3ia5i5wpxfmiyh9x6ibm2ab3-liburcu-aarch64-unknown-linux-gnu-0.14.0/lib/liburcu.so when searching for -lurcu
>        >           /nix/store/wwfrj9kvfi14xclc38qfwm71ah6aawdh-binutils-2.41/bin/ld: skipping incompatible /nix/store/sr2520ilxww86sqscwsj7cynh38zyvxf-keyutils-aarch64-unknown-linux-gnu-1.6.3-lib/lib/libkeyutils.so when searching for -lkeyutils
>        >           /nix/store/wwfrj9kvfi14xclc38qfwm71ah6aawdh-binutils-2.41/bin/ld: skipping incompatible /nix/store/sr2520ilxww86sqscwsj7cynh38zyvxf-keyutils-aarch64-unknown-linux-gnu-1.6.3-lib/lib/libkeyutils.a when searching for -lkeyutils
>        >           /nix/store/wwfrj9kvfi14xclc38qfwm71ah6aawdh-binutils-2.41/bin/ld: cannot find -lkeyutils: No such file or directory
>        >           /nix/store/wwfrj9kvfi14xclc38qfwm71ah6aawdh-binutils-2.41/bin/ld: skipping incompatible /nix/store/sr2520ilxww86sqscwsj7cynh38zyvxf-keyutils-aarch64-unknown-linux-gnu-1.6.3-lib/lib/libkeyutils.so when searching for -lkeyutils
>        >           collect2: error: ld returned 1 exit status
> ```
 
`binutils.bintools` will be for/target the wrong system

Something has to be substituted in the source because ar is hardcoded

--- a/pkgs/by-name/bc/bcachefs-tools/package.nix
+++ b/pkgs/by-name/bc/bcachefs-tools/package.nix
@@ -21,6 +21,7 @@
   makeWrapper,
   nix-update-script,
   python3,
+  fetchpatch,
   fuseSupport ? false,
 }:
 
@@ -71,6 +72,11 @@ stdenv.mkDerivation (finalAttrs: {
     "INITRAMFS_DIR=${placeholder "out"}/etc/initramfs-tools"
   ] ++ lib.optional fuseSupport "BCACHEFS_FUSE=1";
 
+  env = {
+    CARGO_BUILD_TARGET = stdenv.hostPlatform.rust.rustcTargetSpec;
+    "CARGO_TARGET_${stdenv.hostPlatform.rust.cargoEnvVarTarget}_LINKER" = "${stdenv.cc.targetPrefix}cc";
+  };
+
   # FIXME: Try enabling this once the default linux kernel is at least 6.7
   doCheck = false; # needs bcachefs module loaded on builder
 
@@ -78,7 +84,18 @@ stdenv.mkDerivation (finalAttrs: {
     # code refactoring of bcachefs-tools broke reading passphrases from stdin (vs. terminal)
     # upstream issue https://github.com/koverstreet/bcachefs-tools/issues/261
     ./fix-encrypted-boot.patch
+    # https://github.com/koverstreet/bcachefs-tools/pull/305
+    (fetchpatch {
+      name = "use-ar-var-in-makefile.patch";
+      url = "https://github.com/koverstreet/bcachefs-tools/commit/91e67ab2bd48fa135a1f5109b23899a4f1019a03.patch";
+      sha256 = "sha256-nB4Tgcwa8eeasIDQ4rrYORie/X8LMuCSRi+WJNw+R/U=";
+    })
   ];
+
+  postPatch = ''
+    substituteInPlace Makefile \
+      --replace-fail "target/release/bcachefs" "target/${stdenv.hostPlatform.rust.rustcTargetSpec}/release/bcachefs"
+  '';
 
   preCheck = lib.optionalString (!fuseSupport) ''
     rm tests/test_fuse.py

